### PR TITLE
Closes #2405: Add server <-> server transfers of pdarrays

### LINKED ISSUE #2405: Add server <-> server transfers of pdarrays
Add the ability to send a pdarray from one server to
another server through ZMQ sockets. This can be useful when
multiple Arkouda servers would like to operate on smaller
subsets of one giant dataset that can't fit on each Arkouda
server individually. For example, one large "mother" Arkouda
server could load in the entire dataset, down-select the data
as needed for each "child" server, then send the portion of
that array to the child server that would like to operate on
that portion of the dataset.

Prior to this PR, this would need to be done by writing the
down-selected portion from the mother server, reading those
files into the child server, and then operating. This aims
to replace the file system go-between of this use case,
simplifying the number of steps needed to get a pdarray
from one server to another and also potentially achieving 
better performance than would be possible using the file
system.

Usage:

Server-1 ("mother" server):
```python3
>>> a = ak.read('my_file') # file that's too big for other servers
>>> b = a[0:int((a.size/2))] # get a portion of the array to send to child
>>> b.send_array("child-name", 1234) # send smaller array to server running on "child-name"
```

Server-2 ("child" server):
```python3
a = ak.receive_array("mother-name", 1234) # receive array from "mother-name"
                                    # port must match the port from the `send` call
```

Types supported in initial implementation with `send_array`/`receive_array` interface:
- all native Arkouda datatypes (int, uint, bool, float)
- categorical
- strings
- segmented array

Additionally, DataFrames containing columns of the above supported types can be transferred with a slightly different interface of `DataFrame.transfer(hostname, port)`/`ak.receive_dataframe(hostname, port)`

### PR DESCRIPTION
**Pull request by bmcdonald3**
This PR adds the ability to send a pdarray from one server to
another server through ZMQ sockets. This can be useful when
multiple Arkouda servers would like to operate on smaller
subsets of one giant dataset that can't fit on each Arkouda
server individually. For example, one large "mother" Arkouda
server could load in the entire dataset, downselect the data
as needed for each "child" server, then send the portion of
that array to the child server that would like to operate on
that portion of the dataset.

Prior to this PR, this would need to be done by writing the
downselected portion from the mother server, reading those
files into the child server, and then operating. This aims
to replace the file system go-between of this use case,
simplifying the number of steps needed to get a pdarray
from one server to another and also potentially achieving 
better performance than would be possible using the file
system.

The performance could be improved, especially for the case
where the sending/receiving servers are of different element
sizes, but this initial work is focused on correctness.

Performance numbers collected on a Cray CS HDR transfers over
100G HDR-IB:

| Nodes     | Receiving 2  | Receiving 3  | Receiving 4  |
| --------: | -----------: | -----------: | -----------: |
| Sending 2 | 1.2385 GiB/s | 0.6264 GiB/s | 1.4077 GiB/s |
| Sending 3 | 0.6680 GiB/s | 1.6672 GiB/s | 0.6270 GiB/s |
| Sending 4 | 1.3586 GiB/s | 0.7278 GiB/s | 1.9504 GiB/s |

Types supported in initial implementation with `send_array`/`receive_array` interface:
- all native Arkouda datatypes (int, uint, bool, float)
- categorical
- strings
- segmented array

Additionally, DataFrames containing columns of the above supported types can be transferred with a slightly different interface of `DataFrame.transfer(hostname, port)`/`ak.receive_dataframe(hostname, port)`

Closes: #2405

### CODE REVIEW

**Comment by hokiegeek2** [comment]
@bmcdonald3 this looks cool. Question -> when an array is being transferred from one arkouda instance to another, the zmq  ports of the sending and receive servers block until the array transfer completes, correct?

**Comment by bmcdonald3** [comment]
@hokiegeek2 Yes, that is correct

**Comment by bmcdonald3** [comment]
@hokiegeek2 No, I don't have that information, but, based on the numbers above, we observed 0.1 GiB/s per-node as-is, and that was without targeting the high speed network (as mentioned, this code is unoptimized and there are clear next steps for optimization). If we assume this 50 TB array is over 50 nodes, that would be 5 GiB/s injection, so about 15 minutes.

But, aside from the raw numbers of this initial implementation, what I would say in response to that  is that, since this aims to replace writing/reading files (the only way to accomplish moving data from one server to another today), if we are targeting the high speed network (as indicated, the numbers in the table in the original PR are not, but saw large speedups when targeting infiniband), we have a much higher potential injection rate than doing so through the filesystem. If we assume 100 GiB/s infiniband (actual number) as the high bound, and 25 GiB/s to read/write files as a high bound (optimistic numbers), that would mean this would go 8x faster than reading/writing the files (since with files we have to read and write, the injection rate would have to take into both the writing and reading stages, so the injection rate here from reading/writing is actually 12.5 GiB/s, even though we can read/write at 25 GiB/s, so 8x lower than the 100 GiB/s potential from transferring via sockets).

So, while this would block the server for a while when transferring a large array, I don't see that as unique to this operation and think that we would see significant improvements from transferring the files as opposed to reading/writing the files in terms of performance (which is equivalent to the amount of unresponsive time you mention).

**Comment by hokiegeek2** [comment]
@bmcdonald3 gotcha, okay. Just curious, what kind of network did you test on?

**Comment by bmcdonald3** [comment]
Good question! This was tested using 1G Ethernet (I've updated the original description to say that).

In practice, we wouldn't want to use Ethernet for these transfers in most cases, so those performance numbers aren't particularly valuable I don't think and they were really collected as a correctness test more than for the performance and Elliot advised against putting out that performance table, but the table looked nice and I had already collected the numbers, so I had to include it haha.

I plan to run some actual studies using IPoIB in the near future to test more realistic numbers, but since this initial implementation was focused on establishing a baseline with correctness, rather than performance, I have not had a chance to do a full sweep of that yet.

**Comment by pierce314159** [comment]
> @hokiegeek2 No, I don't have that information, but, based on the numbers above, we observed 0.1 GiB/s per-node as-is, and that was without targeting the high speed network (as mentioned, this code is unoptimized and there are clear next steps for optimization). If we assume this 50 TB array is over 50 nodes, that would be 5 GiB/s injection, so about 15 minutes.
> 
> But, aside from the raw numbers of this initial implementation, what I would say in response to that is that, since this aims to replace writing/reading files (the only way to accomplish moving data from one server to another today), if we are targeting the high speed network (as indicated, the numbers in the table in the original PR are not, but saw large speedups when targeting infiniband), we have a much higher potential injection rate than doing so through the filesystem. If we assume 100 GiB/s infiniband (actual number) as the high bound, and 25 GiB/s to read/write files as a high bound (optimistic numbers), that would mean this would go 8x faster than reading/writing the files (since with files we have to read and write, the injection rate would have to take into both the writing and reading stages, so the injection rate here from reading/writing is actually 12.5 GiB/s, even though we can read/write at 25 GiB/s, so 8x lower than the 100 GiB/s potential from transferring via sockets).
> 
> So, while this would block the server for a while when transferring a large array, I don't see that as unique to this operation and think that we would see significant improvements from transferring the files as opposed to reading/writing the files in terms of performance (which is equivalent to the amount of unresponsive time you mention).

@bmcdonald3 and @hokiegeek2 i gotta question. I think this is really interesting, im curious how this would compare in the case when you have multiple "child" servers? So if everyone on a team of `N` people wanted to play with their own copy of the filtered down dataset. With the transfer approach, we'd need to do `N` transfers (one for each "child" server). With the file system approach, we'd only have to do a single write and do `N` reads. Also (my ignorance of file IO is showing :) ) would it be possible for those reads to be done simultaneously? Because if so, it feels like for a large enough number of team-members the file system would be faster right? It's possible that number is too high to every realistically hit and it's possible that none of what i said made any sense lol

**Comment by pierce314159** [comment]
> Elliot advised against putting out that performance table, but the table looked nice and I had already collected the numbers, so I had to include it haha.
> 
lmao the table _does_ look nice 😁

**Comment by bmcdonald3** [comment]
> im curious how this would compare in the case when you have multiple "child" servers? So if everyone on a team of N people wanted to play with their own copy of the filtered down dataset. With the transfer approach, we'd need to do N transfers (one for each "child" server). With the file system approach, we'd only have to do a single write and do N reads. Also (my ignorance of file IO is showing :) ) would it be possible for those reads to be done simultaneously?

This is interesting question, but I think there are a lot of unknowns and this gets into a classic networking problem. If every single server did want the exact same dataset, then I think the most efficient way to do that would be to have the head server transfer to a child, then have that child become another worker node, so that child would then transfer it's data to another server, so you'd get something like: 1 node -> transfer -> 2 nodes -> transfer -> 4 nodes -> transfer -> 8 nodes ..., where each new child node is starting to do some transfers as well, so I think you could get `log(N)` time there.

This would definitely be a best case scenario though, it would be pretty hard to get a bunch of different people acting in unison to make something like that happen, but if the goal was just freeing up the one big server, then that one could just do one transfer and then have that child be the new server to farm things out, but all of this is speculative, since I don't have a concrete use case in mind that I know this is looking to help with, it is possible that IO is better, but it would depend on the use case.

What I was imagining based on past conversations was that every child server wanted a different portion of the dataset, which is a very different problem.

This is information that I stole from Elliot, so I can't say a ton about it, but networks and file systems typically have three different limits: limit for single thread transfer, limit for single node transfer, limit for total simultaneous transfer amongst all nodes. Inifiniband has a much higher total simultaneous transfer upper bound than the typical file system. So if we were really working with huge 50TB arrays like John mentioned, we would definitely be hitting that third bottleneck of total simultaneous transfer amongst all nodes, where if we were trying to read 50 TB in from the same filesystem on 5 different servers, that will basically be serialized and all the servers would be doing their reads one at a time, not in parallel. On much smaller data sets where you aren't capping out that total transfer rate on a single server, the parallelism from reading from multiple servers could come in handy, but that wouldn't be a case where the server was stalling for 15 minutes.

In summary, I am not an expert on this stuff and I think that the real value here is dependent on the actual use case in question.

**Comment by pierce314159** [comment]
> This is interesting question, but I think there are a lot of unknowns and this gets into a classic networking problem. If every single server did want the exact same dataset, then I think the most efficient way to do that would be to have the head server transfer to a child, then have that child become another worker node, so that child would then transfer it's data to another server, so you'd get something like: 1 node -> transfer -> 2 nodes -> transfer -> 4 nodes -> transfer -> 8 nodes ..., where each new child node is starting to do some transfers as well, so I think you could get `log(N)` time there.
> 
Ooooo i didn't think about using the child servers to do transfers, that's clever
> if we were trying to read 50 TB in from the same filesystem on 5 different servers, that will basically be serialized and all the servers would be doing their reads one at a time, not in parallel. On much smaller data sets where you aren't capping out that total transfer rate on a single server, the parallelism from reading from multiple servers could come in handy, but that wouldn't be a case where the server was stalling for 15 minutes.
> 
Okay that makes a lot of sense! thanks ben/elliot

File: arkouda/io.py [hunk 1]
```diff
@@ -1912,3 +1913,43 @@ def read_tagged_data(
...
+        The hostname of the pdarray that sent the array
+    port : int_scalars
+        The port to send the array over. This needs to be an
+        open port (i.e., not one that the Arkouda server is
+        running on). This will open up `numLocales` ports,
+        each of which in succession, so will use ports of the
+        range {port..(port+numLocales)} (e.g., running an
+        Arkouda server of 4 nodes, port 1234 is passed as
+        `port`, Arkouda will use ports 1234, 1235, 1236,
+        and 1237 to send the array data).
```
Comment (pierce314159) [inline arkouda/io.py:1934]: would these ports need to be the same or distinct from those in `send_array`? I'm guessing distinct (or at least not necessarily the same) because we could have a different number of locales between the receiving and sending servers right?

File: arkouda/categorical.py [hunk 2]
```diff
@@ -1356,3 +1359,51 @@ def parse_hdf_categoricals(
...
+            receive the Categorical is running.
+        port : int_scalars
+            The port to send the array over. This needs to be an
+            open port (i.e., not one that the Arkouda server is
+            running on). This will open up `numLocales` ports,
+            each of which in succession, so will use ports of the
+            range {port..(port+numLocales)} (e.g., running an
+            Arkouda server of 4 nodes, port 1234 is passed as
+            `port`, Arkouda will use ports 1234, 1235, 1236,
+            and 1237 to send the array data).
```
Comment (pierce314159) [inline arkouda/categorical.py:1380]: i don't know much about networking but i wanted to throw this out there. Idk if it's a reasonable assumption that a block of num locales ports would be free. I know that num_locales can get pretty big in real use cases, so I just wanted to check that this was something ppl who know things about networks had thought about lol

**Review by pierce314159 — COMMENTED** [review]
my initial review, I got a couple questions mostly me not understanding ports and stuff that well

I didn't really dig into the chpl side too much. I figured i would let @hokiegeek2 and @Ethan-DeBandi99 take a crack at that first since they have more expertise. i'll take a look on my next pass through

File: arkouda/categorical.py [hunk 3]
```diff
@@ -200,7 +200,10 @@ def from_return_msg(cls, rep_msg) -> Categorical:
         eles = json.loads(rep_msg)
         codes = create_pdarray(eles["codes"])
         cats = Strings.from_return_msg(eles["categories"])
-        na_code = create_pdarray(eles["_akNAcode"])
+        if "_akNAcode" in eles:
+            na_code = create_pdarray(eles["_akNAcode"])
+        else:
+            na_code = None
```
Comment (Ethan-DeBandi99) [inline arkouda/categorical.py:206]: `ak.Categorical._akNAcode` should always be defined. Looking at the `init`, there are cases where it will not be a `pdarray` (which means we there are alterations we need to make unrelated to this PR). However, in this case, `_akNAcode` should always be defined and should always be a `pdarray`.

File: arkouda/categorical.py [hunk 4]
```diff
@@ -1356,3 +1359,51 @@ def parse_hdf_categoricals(
...
+            Raised if the op is not within the pdarray.BinOps set
+        TypeError
+            Raised if other is not a pdarray or the pdarray.dtype is not
+            a supported dtype
+        """
+        # hostname is the hostname to send to
+        args = {
+            "codes": self.codes,
+            "categories": self.categories,
+            "objType": "categorical",
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/categorical.py:1401]: I agree with @pierce314159 that we should be using `self.objType`, but disagree that it can wait. We really should be doing this with all new code sending the object type to prevent adding additional tech debt.

File: arkouda/dataframe.py [hunk 5]
```diff
@@ -641,6 +642,76 @@ def _get_head_tail_server(self):
         new_df._set_index(self.index.index[idx])
         return new_df.to_pandas(retain_index=True)[self._columns]
 
+    def transfer(self, hostname, port):
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/dataframe.py:645]: Just noting that I agree with @pierce314159 here.

File: arkouda/categorical.py [hunk 6]
```diff
@@ -1356,3 +1359,51 @@ def parse_hdf_categoricals(
...
+            "NA_codes": self._akNAcode,
```
Comment (Ethan-DeBandi99) [inline arkouda/categorical.py:1402]: Based on the `__init__`, it looks like there are cases where this can be an `int`. Maybe I just haven't seen it yet, but are we checking that somewhere? If not, it would probably be best to create the `pdarray` here.

Again, this is definitely not checked in other locations in existing code. I will add an issue to look into this and make sure it gets updated.

File: arkouda/dataframe.py [hunk 7]
```diff
@@ -641,6 +642,76 @@ def _get_head_tail_server(self):
...
+            elif isinstance(self[col], Fields):
+                msg_list.append(f"Fields+{col}+{self[col].name}")
+            elif isinstance(self[col], IPv4):
+                msg_list.append(f"IPv4+{col}+{self[col].name}")
+            elif isinstance(self[col], Datetime):
+                msg_list.append(f"Datetime+{col}+{self[col].name}")
+            elif isinstance(self[col], BitVector):
+                msg_list.append(f"BitVector+{col}+{self[col].name}")
+            else:
+                msg_list.append(f"pdarray+{col}+{self[col].name}")
```
Comment (Ethan-DeBandi99) [inline arkouda/dataframe.py:698]: We are trying to move away from delimited strings within the messaging layer of Arkouda. It looks like it would be a rather large up, but for what it is worth, we typically use multiple lists in these cases where the indexes line everything up. Additionally, that would allow for using the `objType` properties and the `ObjType` enum on the server for consistency. We could also do that here, but again I think we want to move away from delimited strings for messages. 

Also, noting that to that some classes/objects may need to have the `objType` property added in the class and to the enum on the server.

File: src/TransferMsg.chpl [hunk 8]
```diff
@@ -0,0 +1,883 @@
...
+
+      var localeCount = receiveLocaleCount(hostname, port:string);
+      
+      sendDataFrameSetupInfo(port:string, numColumns, eleList: string);
+
+      var colNames = "";
+      for ele in eleList {
+        var ele_parts = ele.split("+");
+        sendColumnName(port, ele_parts[1]);
+        if ele_parts[0] == "Categorical" {
```
Comment (Ethan-DeBandi99) [inline src/TransferMsg.chpl:62]: It would be good if we use the `ObjType` enum here. I think we will need to add some things to it for that.

File: src/TransferMsg.chpl
Comment (Ethan-DeBandi99) [inline src/TransferMsg.chpl:1]: I have just a general implementation question. This may be due to a lack of experience with ZMQ, but just want to clarify. In general, all ZMQ traffic in Arkouda should be going through locale0. It looks like in this implementation, each Locale is getting its own port to communicate on, but doesn't that traffic still go through locale0? I know there have been previous things that we decided not to do because even opening a separate port would still block the existing an vice versa. I am just worried that if that is the case, even if multiple locales are sending or receiving at once they will be blocking each other. Maybe I am completely missing something, or just need to be reminded of something I've already had explained to me. If that is the case I apologize, but could we get clarification on this?

**Review by Ethan-DeBandi99 — COMMENTED** [review]
Couple of things to address. And one question for clarification.

File: src/TransferMsg.chpl [hunk 9]
```diff
@@ -0,0 +1,883 @@
...
+    use Map;
+    use ServerErrors;
+    use ServerConfig;
+    use ServerErrorStrings;
+
+    use SegmentedString;
+
+    proc sendDataFrameSetupInfo(port:string, numColumns: int, elements: string) throws {
+      var context: Context;
+      var socket = context.socket(ZMQ.PUSH);
```
[the remark this answers is not in the document]
Comment (bmcdonald3) [reply src/TransferMsg.chpl:20]: This transfer will happen from locale0 on the sending server to locale0 on the receiving server and will only use 1 port. This transfer only happens once per dataframe transfer (see line 56).

File: arkouda/io.py (hunk 1 above)
Comment (bmcdonald3) [reply arkouda/io.py:1934]: It will be the number of ports from the server with more locales, so, if we had a server sending with 4 locales and only receiving on 2, both of the servers will open up the 4 ports (since those are the ports that were sent on).

File: src/TransferMsg.chpl
Comment (bmcdonald3) [reply src/TransferMsg.chpl:1]: > all ZMQ traffic in Arkouda should be going through locale0

Is there a reason this should be the case? I think this is the case for the usage of Arkouda's ZMQ today, but that is not the case with how it is used in this PR. 

> doesn't that traffic still go through locale0

No, each locale has it's own ZMQ instance that is handling the traffic.

I'm not sure if I'm answering the question here, but each separate locale is doing it's own communication, so there won't be any overlap with the existing usage of ZMQ for message sending or message sending from the different locales (which is why you see the performance get better as you increase the number of locales in the initial post). Does that help?

File: arkouda/categorical.py (hunk 3 above)
Comment (bmcdonald3) [reply arkouda/categorical.py:206]: OK I think I did something stupid there, working to get that fixed now, thanks for pointing out.

File: arkouda/io.py (hunk 1 above)
Comment (pierce314159) [reply arkouda/io.py:1934]: cool thanks for clarifying!

File: arkouda/dataframe.py (hunk 7 above)
Comment (Ethan-DeBandi99) [reply arkouda/dataframe.py]: @bmcdonald3 - I added a note to issue #2395 to make sure that when the code this is modeled after is updated to move away from delimited. I am not thrilled about additional tech debt here, but I think in this instance I am ok with it.

File: src/TransferMsg.chpl
Comment (Ethan-DeBandi99) [reply src/TransferMsg.chpl:1]: @bmcdonald3 - I think so. I guess my only other question would be do the sockets on each locale block one another? I am assuming not based on the performance, but would rather do the sanity check now.

Comment (bmcdonald3) [reply src/TransferMsg.chpl:1]: No, they don't. ZMQ messages are sent asynchronously in a non-blocking manner, but there are additional dials in ZMQ you can change to modify that behavior. For the Chapel ZMQ usage, we have it default to use `ZMQ_DONTWAIT`, which is their way of saying non-blocking.

Comment (Ethan-DeBandi99) [reply src/TransferMsg.chpl:1]: Perfect thanks.

**Review by pierce314159 — APPROVED** [review]

**Comment by bradcray** [comment]
Sweet!
